mtp_api/apps/security/models.py

```diff
@@ -0,0 +1,97 @@
+from functools import reduce
+from itertools import chain
+
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+from prison.models import Prison
+
+
+class SenderProfile(TimeStampedModel):
+    credit_count = models.IntegerField(default=0)
+    credit_total = models.IntegerField(default=0)
+
+    @property
+    def credit_filters(self):
+        return (
+            reduce(
+                lambda x, y: x | y,
+                chain(
+                    (
+                        models.Q(transaction__sender_name=d.sender_name) &
+                        models.Q(transaction__sender_sort_code=d.sender_sort_code) &
+                        models.Q(transaction__sender_account_number=d.sender_account_number) &
+                        models.Q(transaction__sender_roll_number=d.sender_roll_number)
+                        for d in self.bank_transfer_details.all()
+                    ),
+                    (
+                        models.Q(payment__card_number_last_digits=d.card_number_last_digits) &
+                        models.Q(payment__card_expiry_date=d.card_expiry_date)
+                        for d in self.debit_card_details.all()
+                    )
+                ),
+                models.Q(pk=None)
+            )
+        )
+
+    class Meta:
+        permissions = (
+            ('view_senderprofile', 'Can view sender profile'),
+        )
+
+
+class BankTransferSenderDetails(TimeStampedModel):
+    sender_name = models.CharField(max_length=250, blank=True)
+    sender_sort_code = models.CharField(max_length=50, blank=True)
+    sender_account_number = models.CharField(max_length=50, blank=True)
+    sender_roll_number = models.CharField(max_length=50, blank=True)
+    sender = models.ForeignKey(
+        SenderProfile, on_delete=models.CASCADE, related_name='bank_transfer_details'
+    )
+
```

Comment:
i think it should be, because e.g. then we can count metro bank senders separately because although they all have the same sort code/account number they do have different sender names